iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/transformation/datastructure/row/SerializableRowList.java

```diff
@@ -0,0 +1,365 @@
+    SerializationRecorder recorder = new SerializationRecorder(queryId);
+    return new SerializableRowList(recorder, dataTypes);
+  }
+
+  private SerializableRowList(SerializationRecorder serializationRecorder, TSDataType[] dataTypes) {
+    this.serializationRecorder = serializationRecorder;
+    this.dataTypes = dataTypes;
+    serde = new TsBlockSerde();
+    blockSizes = new ArrayList<>();
+
+    valueColumnCount = dataTypes.length;
+    prefixNullCount = 0;
+    skipPrefixNullCount = 0;
+
+    init();
+  }
+
+  /**
+   * Calculate the number of rows that can be cached given the memory limit.
+   *
+   * @param dataTypes Data types of columns.
+   * @param memoryLimitInMB Memory limit.
+   * @param byteArrayLengthForMemoryControl Max memory usage for a {@link TSDataType#TEXT}.
+   * @return Number of rows that can be cached.
+   * @throws QueryProcessException if the result capacity <= 0
+   * @throws UnSupportedDataTypeException if the input datatype can not be handled by the given
+   *     branches.
+   */
+  protected static int calculateCapacity(
+      TSDataType[] dataTypes, float memoryLimitInMB, int byteArrayLengthForMemoryControl)
+      throws QueryProcessException {
+    int rowLength = ReadWriteIOUtils.LONG_LEN; // timestamp
+    for (TSDataType dataType : dataTypes) { // fields
+      switch (dataType) {
+        case INT32:
+          rowLength += ReadWriteIOUtils.INT_LEN;
+          break;
+        case INT64:
+          rowLength += ReadWriteIOUtils.LONG_LEN;
+          break;
+        case FLOAT:
+          rowLength += ReadWriteIOUtils.FLOAT_LEN;
+          break;
+        case DOUBLE:
+          rowLength += ReadWriteIOUtils.DOUBLE_LEN;
+          break;
+        case BOOLEAN:
+          rowLength += ReadWriteIOUtils.BOOLEAN_LEN;
+          break;
+        case TEXT:
+          rowLength +=
+              MIN_OBJECT_HEADER_SIZE + MIN_ARRAY_HEADER_SIZE + byteArrayLengthForMemoryControl;
+          break;
+        default:
+          throw new UnSupportedDataTypeException(dataType.toString());
+      }
+    }
+    rowLength += ReadWriteIOUtils.BIT_LEN; // null field
+
+    int capacity = (int) (memoryLimitInMB * MB / 2 / rowLength);
+    if (capacity <= 0) {
+      throw new QueryProcessException("Memory is not enough for current query.");
+    }
+    return capacity;
+  }
+
+  public int size() {
+    return prefixNullCount + skipPrefixNullCount;
+  }
+
+  public int getBlockCount() {
+    // Threat prefix null values as one column
+    int additional_null_block = prefixNullCount == 0 ? 0 : 1;
+    return blockSizes.size() + additional_null_block;
+  }
+
+  public int getBlockSize(int index) {
+    if (prefixNullCount != 0) {
+      return index == 0 ? prefixNullCount : blockSizes.get(index - 1);
+    } else {
+      return blockSizes.get(index);
+    }
+  }
+
+  // region single data methods
+  public long getTime(int index) {
+    // Never access null row's time
+    assert index >= prefixNullCount;
```

Comment:
replace assert with checkState()

Reply:
What are the benefits of the `checkState()` over the `asset`?

Reply:
Done.